src/libxtp/segment.cc

```diff
@@ -0,0 +1,542 @@
+
+  return (state == -1) ? _U_cC_nN_e : _U_cC_nN_h;
+}
+
+const double &Segment::getU_nC_nN(int state) {
+
+  return (state == -1) ? _U_nC_nN_e : _U_nC_nN_h;
+}
+
+const double &Segment::getU_cN_cC(int state) {
+
+  return (state == -1) ? _U_cN_cC_e : _U_cN_cC_h;
+}
+
+double Segment::getSiteEnergy(int state) {
+
+  double result;
+  if (state == -1) {
+    result = getEMpoles(state) + _U_cC_nN_e;
+  } else if (state == +1) {
+    result = getEMpoles(state) + _U_cC_nN_h;
+  } else if (state == +2) {
+    result = getEMpoles(state) + _U_xX_nN_s;
+  } else if (state == +3) {
+    result = getEMpoles(state) + _U_xX_nN_t;
+  } else {
+    throw runtime_error(
+        " ERROR CODE whe__00s11o__");  // blabla what do I do here?
+  }
+  return result;
+}
+
+void Segment::setEMpoles(int state, double energy) {
+
+  _hasChrgState.resize(5);
+  _hasChrgState[state + 1] = true;
+  _eMpoles[state + 1] = energy;
+}
+
+double Segment::getEMpoles(int state) {
+
+  return _eMpoles[state + 1] - _eMpoles[1];
+}
+
+void Segment::AddFragment(Fragment *fragment) {
+  _fragments.push_back(fragment);
+  fragment->setSegment(this);
+}
+
+void Segment::AddAtom(Atom *atom) {
+  _atoms.push_back(atom);
+  atom->setSegment(this);
+}
+
+void Segment::AddPolarSite(PolarSite *pole) {
+
+  _polarSites.push_back(pole);
+  pole->setSegment(this);
+}
+
+void Segment::AddAPolarSite(APolarSite *pole) {
+
+  _apolarSites.push_back(pole);
+  pole->setSegment(this);
+}
+
+void Segment::calcPos() {
+  vec pos = vec(0, 0, 0);
+  double totWeight = 0.0;
+
+  for (unsigned int i = 0; i < _atoms.size(); i++) {
+    pos += _atoms[i]->getPos() * _atoms[i]->getWeight();
+    totWeight += _atoms[i]->getWeight();
+  }
+
+  _CoM = pos / totWeight;
+}
+
+void Segment::calcApproxSize() {
+  _approxsize = 0.0;
+
+  tools::vec min = vec(numeric_limits<double>::max());
+  tools::vec max = vec(numeric_limits<double>::min());
+  vector<Fragment *>::iterator fragit1;
+  for (fragit1 = Fragments().begin(); fragit1 < Fragments().end(); fragit1++) {
+    const tools::vec &pos = (*fragit1)->getPos();
+    if (pos.getX() > max.getX()) {
+      max.x() = pos.getX();
+    } else if (pos.getX() < min.getX()) {
+      min.x() = pos.getX();
+    }
+    if (pos.getY() > max.getY()) {
+      max.y() = pos.getY();
+    } else if (pos.getY() < min.getY()) {
+      min.y() = pos.getY();
+    }
+    if (pos.getZ() > max.getZ()) {
+      max.z() = pos.getZ();
+    } else if (pos.getZ() < min.getZ()) {
+      min.z() = pos.getZ();
+    }
+  }
+
+  _approxsize = abs(max - min);
+
+  return;
+}
+
+void Segment::Rigidify() {
+
+  if (this->getType()->canRigidify()) {
+    // Establish which atoms to use to define local frame
+    vector<Fragment *>::iterator fit;
+
+    for (fit = this->Fragments().begin(); fit < this->Fragments().end();
+         fit++) {
+      (*fit)->Rigidify();
+    }
+  } else {
+    return;
+  }
+}
+
+void Segment::WritePDB(FILE *out, string tag1, string tag2) {
```

Comment:
another pdb writer

Reply:
Right!

Reply:
I can consolidate this with the pdb reader and writer in csg if this is useful for this class to have them e.g. for testing or whatever. Or if it is not useful we can just get rid of it.

Reply:
Maybe get rid of them for now and relink to the csg readers when needed later.

Reply:
Fair enough.

Reply:
I personally think the file readers and writers should allow every atom container to write and read files.